### DISCUSSION TITLE
Using API with next export

### DISCUSSION THREAD

**Discussion by robhrt7** [vercel/vercel#6551] [Help] [answered]
When building a statically exported website with Next.js, hosted on Vercel, we're seeing this message in the build log:

```
warn  - Statically exporting a Next.js application via `next export` disables API routes.
```

It's not clear though, going trhough the docs, how to re-enable API routes and how is this limitation justified. We build a fully pre-rendered website connected to a Headless eCom platform, we don't need any dynamic routes (eg get server side props), but we do need API for our eCom tasks that we would like to host on the same project. Is there really no way to achieve this? 

We'd like to keep our website using static export, as it's the most performance option, in such case we need to maintain another Vercel project only for API (serverless function)?

**Comment by leerob** [comment, marked as answer]
Correct, you can't use API routes with a completely static export of `next export`. 

To re-enable API routes, don't use `next export` and just use `next build` 😄 Next.js will automatically optimize for static on Vercel, plus you can use API routes.

**Comment by robhrt7** [comment]
In this particular use case we build a fully static website and we’re using exportPathMap, that’s is not working with `next build`. We found it impossible to use `getStaticPaths`, when building websites that are fully controlled by a Headless CMS, with an ability of creating pages with any URL plus having some other requirements on url building like localization (eg `en-gn/url`).

It’s not clear what technically restricts using API routes with a fully static website, when the APIs could be used exclusively for calls from client side, or for some eCom operations. So I believe the only workaround is to use two projects on Vercel?

**Reply by leerob** [reply]
Could you expand more on the issues you saw with `getStaticPaths` and clarify the type of application you're building?

**Reply by AlexHramovich** [reply]
Hey, Lee!
Here you can find some details.

**A few words about the type of application:**
- Fully static website
- Fully controlled by the CMS
- The ability to create new pages with any path we need (/details/details/5/... or /about us)
- On the Next.js lvl we have only one page that gets parameters from the query provided by `getInitialProps` and determines how to build it and which template should be used

**How it works**
We use "exportPathMap" to fetch all objects from the CMS and prepare a path map. The result of this function looks like this


```
return cmsPageObjects.reduce((result, cur) => {
  const { id, pagePath, ...rest } = cur;

  return {
    ...result,
    [pagePath]: { page: 'dynamic', query: { cmsObjectId: id, ...rest } },
  };
}, {});
```

Based on the exported paths, we build all the pages through one page named `dynamic.js`

```
DynamicPage.getInitialProps = async ({ query }) => {
  const { cmsObjectId, pageType } = query;
  const { page, config } = await fetchPageData(pageType, cmsObjectId);

  return {
    data: { page, config },
  };
};
```
This gives us the opportunity to be really flexible and do whatever we want.
That's it :)

**Questions and notes:**
1. We cannot use `getStaticProps` in this case, because this function does not provide a `query` based on which we can fetch information about the page from an external resource like CMS.
2. 'getStaticProps' has its own `parameters`, but they are based on the file name, and it is not clear how to handle paths with different levels of nesting ( / .. / .. / .. / .. or /..)
3. Unfortunately, there are no `locale` fields in `getInitialProps`, but it would be great to have this magic
4. Is it correct that the `revalidate` and other features associated with the cache do not work when we use the `next export`, and is it possible to enable it ?

Please let us know if everything is clear now, or in case you need any other details

**Reply by leerob** [reply]
1. I would recommend fetching data from your CMS based on the page routes, not the page query parameters. You can then use these page paths (as a dynamic route) to fetch the correct data from your CMS.
2. This is supported, please read [Dynamic Routing](https://nextjs.org/docs/routing/dynamic-routes) or look at one of the `cms-*` examples in the `examples/` folder.
3. This is supported with `getStaticProps` and using `next start` instead of `next export`
4. This is supported with `getStaticProps` and using `next start` instead of `next export`

Overall, a very small percentage of Next.js applications are using `next export`, and features like i18n routing and ISR require `next start` to work properly. I would recommend using that instead of `next export`, and you won't need to use `exportPathMap` at all, too.

**Reply by AlexHramovich** [reply]
Thanks for the answer! That sounds reasonable.

I created a new project to test how it works in our case.
And found that yes, it is really easy to manage various routes through `[[...slug]].js`.

But it seems that I need one more of advice from you.
How can I provide some more props through the `getStaticPaths` ?
For example, I need to get the 'documentId' in "getStaticProps", but I don't want to see this parameter in the URL.

This can happen if I don't have the ability to fetch a page data using the path, but only through the ID.
And at the same time, I don't want to see this ID in the URL.

**Reply by AlexHramovich** [reply]
@leerob Could you take a look at the question above?
